pydantic_ai_slim/pydantic_ai/models/openai.py

```diff
@@ -160,6 +175,78 @@
 }
 
 
+@dataclass
+class _OpenAIToolChoiceResult:
+    """Intermediate representation of resolved tool choice for OpenAI APIs."""
+
+    kind: Literal['literal', 'single_tool', 'allowed_tools']
+    """The kind of tool choice result."""
+
+    literal_value: Literal['auto', 'required', 'none'] | None = None
+    """For 'literal' kind: the literal value to pass to the API."""
+
+    tool_name: str | None = None
+    """For 'single_tool' kind: the specific tool name."""
+
+    tool_names: list[str] | None = None
+    """For 'allowed_tools' kind: the list of allowed tool names."""
+
+    allowed_mode: Literal['auto', 'required'] = 'required'
+    """For 'allowed_tools' kind: whether the model must use one of the tools or can choose not to."""
+
+
+def _resolve_openai_tool_choice(
+    model_settings: ModelSettings,
+    model_request_parameters: ModelRequestParameters,
+    profile: ModelProfile,
+) -> _OpenAIToolChoiceResult | None:
+    """Resolve tool choice settings into an intermediate representation for OpenAI APIs.
+
+    This centralizes the logic shared between Chat Completions and Responses APIs.
+    Returns None if there are no tools, otherwise returns an _OpenAIToolChoiceResult.
+    """
+    resolved = resolve_tool_choice(model_settings, model_request_parameters)
+    openai_profile = OpenAIModelProfile.from_profile(profile)
+
+    if resolved is None:
+        # Default behavior: infer from allow_text_output
+        if not model_request_parameters.allow_text_output and openai_profile.openai_supports_tool_choice_required:
+            return _OpenAIToolChoiceResult(kind='literal', literal_value='required')
+        return _OpenAIToolChoiceResult(kind='literal', literal_value='auto')
+
+    if resolved.mode == 'auto':
+        return _OpenAIToolChoiceResult(kind='literal', literal_value='auto')
+
+    if resolved.mode == 'required':
+        if not openai_profile.openai_supports_tool_choice_required:
+            warnings.warn(
```

Comment:
- No warning please
- We should be sending a list of function tool names, so the profile setting may not be relevant anyway